airflow/jobs.py

```diff
@@ -547,41 +785,105 @@ def process_dag(self, dag, queue):
+                task_instance.state = new_state
+                session.merge(task_instance)
+        session.commit()
+
+    @provide_session
+    def _execute_task_instances(self,
+                                simple_dag_bag,
+                                states,
+                                session=None):
+        """
+        Fetches task instances from ORM in the specified states, figures
+        out pool limits, and sends them to the executor for execution.
+
+        :param simple_dag_bag: TaskInstances associated with DAGs in the
+        simple_dag_bag will be fetched from the DB and executed
+        :type simple_dag_bag: SimpleDagBag
+        :param executor: the executor that runs task instances
+        :type executor: BaseExecutor
+        :param states: Execute TaskInstances in these states
+        :type states: Tuple[State]
+        :return: None
+        """
+        # Get all the relevant task instances
         TI = models.TaskInstance
-        queued_tis = (
-            session.query(TI)
-            .filter(TI.state == State.QUEUED)
+        task_instances_to_examine = (
+            session
+            .query(TI)
+            .filter(TI.dag_id.in_(simple_dag_bag.dag_ids))
+            .filter(TI.state.in_(states))
             .all()
         )
-        self.logger.info(
-            "Prioritizing {} queued jobs".format(len(queued_tis)))
-        session.expunge_all()
-        d = defaultdict(list)
-        for ti in queued_tis:
-            if ti.dag_id not in dagbag.dags:
-                self.logger.info(
-                    "DAG no longer in dagbag, deleting {}".format(ti))
-                session.delete(ti)
-                session.commit()
-            elif not dagbag.dags[ti.dag_id].has_task(ti.task_id):
-                self.logger.info(
-                    "Task no longer exists, deleting {}".format(ti))
-                session.delete(ti)
-                session.commit()
-            else:
-                d[ti.pool].append(ti)
 
-        dag_blacklist = set(dagbag.paused_dags())
-        for pool, tis in list(d.items()):
+        # Put one task instance on each line
+        if len(task_instances_to_examine) == 0:
+            self.logger.info("No tasks to send to the executor")
+            return
+
+        task_instance_str = "\n\t".join(
+            ["{}".format(x) for x in task_instances_to_examine])
+        self.logger.info("Tasks up for execution:\n\t{}".format(task_instance_str))
+
+        # Get the pool settings
+        pools = {p.pool: p for p in session.query(models.Pool).all()}
+
+        pool_to_task_instances = defaultdict(list)
+        for task_instance in task_instances_to_examine:
+            pool_to_task_instances[task_instance.pool].append(task_instance)
+
+        # Go through each pool, and queue up a task for execution if there are
+        # any open slots in the pool.
+        for pool, task_instances in pool_to_task_instances.items():
```

Comment:
I would expect something like a "pool.select_with_for_update().... commit()" when trying to obtain a slot from the pools. Especially now this has become multiprocess it will definitely lead obtaining more slots from the pool than are actually available (which is already a standing issue)